Product Creation with AI v2 - Milestone 3: progress bar and label with dynamic color and text + logic

pmusolino opened the pull request:
Part of #13274
Closes #13111 and decision-made based on C03L1NF1EA3/p1720512584126239

I'm also waiting for a feedback here C03L1NF1EA3/p1720515544395649

## Description
As part of Milestone 3, I implemented this new view with a progress bar and a label that changes status based on the number of words the user is typing in the starting information screen of Product Creation with AI.
This feature is hidden under a feature flag.

## Changes:
- `ProductCreationAIBarProgressStyle`: created a new ProgressViewStyle that match the style proposed in the Figma design, for the progress bar.
- `ProductCreationAIPromptProgressBar`: the view containing the progress bar, the text and the logic.
- `ProductCreationAIPromptProgressBarViewModel` + unit tests: view model of `ProductCreationAIPromptProgressBar`, containing the logic of the progress bar and the various sttuses. Right now we have 5 statuses matching the Figma design, and the range of words for matching the various statuses.

## Testing information
1. Start creating a new product with AI, and start typing.
2. Check the progress bar color and progress and how the suggestion text changes based on the number of words added.
3. Disable the feature flag `productCreationAIv2M3`, then check that the progress view is hidden.

## Demo
https://github.com/woocommerce/woocommerce-ios/assets/495617/6a4149ea-e54e-4535-baa4-5f9b851520b1



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

pmusolino committed:
feat: initial implementation of Progress bar and label with dynamic color and text

pmusolino committed:
new group PromptProgressBar

pmusolino committed:
feat: implemented `ProductCreationAIPromptProgressBar` inside the Starting Info View of Product Creation with AI

pmusolino committed:
feat: text alignment on the left

pmusolino committed:
feat: ProductCreationAIPromptProgressBarViewModelTests

pmusolino committed:
Merge commit '285181a554340bf7145936e3abb91671c3ca3136' into issue/13118-progress-bar-and-label-with-dynamic-color-and-text

pmusolino committed:
feat: implemented productCreationAIv2M3 feature flag for displaying the progress bar view into Product Creation AI Starting Info View

pmusolino committed:
feat: gray bar by default

pmusolino committed:
fix: space

pmusolino committed:
feat: text animation

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBar.swift:
```diff
@@ -0,0 +1,46 @@
+                Spacer()
+            })
+            .animation(.easeIn, value: viewModel.status)
+            .padding(.top, Layout.padding/2)
```
1. For simplicity, we should specify the spacing in VStack instead of using top padding.
2. The spacing should be `8` instead of `16`.

pmusolino replied:
@selanthiraiyan I strongly followed the padding of the Figma design, where the padding it's 16 from borders, and 8 between the progress bar and the prompt suggestion. 🤔

selanthiraiyan replied:
Here the `padding` modifier with the `top` parameter determines the vertical spacing between the progress view and the text. As per the Figma design, the spacing should be `8`, and currently it is `16`.

<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/a4ba652b-4a85-4f16-871d-bef4afad36dc" width="320"/>

pmusolino replied:
Oh got it. Fixed here 13ea31ad99a745616f85cccafce458d055f3decf

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBar.swift:
```diff
@@ -0,0 +1,46 @@
+        .cornerRadius(Layout.radius)
+        .onChange(of: text, perform: { newText in
+                    viewModel.updateText(to: newText)
+                })
```
Super nit: Indentation adjustment needed.

pmusolino replied:
90274dce9b59a6b9ee103903b47954d993638e39

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:
```diff
@@ -0,0 +1,133 @@
+            case .inProgress:
+                return NSLocalizedString(
+                    "productCreationAIPromptProgressBar.main.inProgress",
+                    value: "Add more details. ",
```
I think we cannot be sure that the empty space at the end of the string will be retained after translations. How about we manually add an empty space `Text` [here](https://github.com/woocommerce/woocommerce-ios/pull/13280/files#diff-dfcce39268903acf8f3334776759bad7d7d85a4dbc0974b2eaecd6cbeef4f036R17). This applies to all other strings in `mainDescription`.

pmusolino replied:
Good suggestion cbc16beeb528a660951e50b6d76dc0998e26a1c8

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:
```diff
@@ -0,0 +1,133 @@
+            case .almostDone:
+                return NSLocalizedString(
+                    "productCreationAIPromptProgressBar.main.almostDone",
+                    value: "Great prompt! ",
```
I think this should be "Almost there.". 

<img width="266" alt="image" src="https://github.com/woocommerce/woocommerce-ios/assets/524475/289e311b-ebdb-4401-a6f0-f65edd8f0193">

pmusolino replied:
Good catch cbc16beeb528a660951e50b6d76dc0998e26a1c8

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:
```diff
@@ -0,0 +1,133 @@
+                return NSLocalizedString(
+                    "productCreationAIPromptProgressBar.main.almostDone",
+                    value: "Great prompt! ",
+                    comment: "State when the prompt description is great for the main prompt description suggestion in product creation with AI."
```
We have to update the comment as well.

pmusolino replied:
6933fae89817f049395aaac016895d73fd6e49eb

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:
```diff
@@ -0,0 +1,133 @@
+            case .almostDone:
+                return NSLocalizedString(
+                    "productCreationAIPromptProgressBar.secondary.almostDone",
+                    value: "Where was it made?",
```
❓ Should we use a more generic text here? We can't be sure that the user hasn't already answered this question in the keywords. What do you think?

pmusolino replied:
I edited it to `Mention additional relevant information or characteristics.`. Wdyt? 998206104e7e9837a9d1d242922697c779c6a6a7

selanthiraiyan replied:
Sounds good to me. 👍

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/ProductCreationAIStartingInfoView.swift:
```diff
@@ -72,6 +72,9 @@ struct ProductCreationAIStartingInfoView: View {
                         )
                     }
 
+                    ProductCreationAIPromptProgressBar(text: $viewModel.features)
+                        .renderedIf(viewModel.featureFlagService.isFeatureFlagEnabled(.productCreationAIv2M3))
```
nit: We could use an if condition to ensure that the progress bar-related code isn't executed at all when the feature flag is off. 

```suggestion
                    if viewModel.featureFlagService.isFeatureFlagEnabled(.productCreationAIv2M3 {
                        ProductCreationAIPromptProgressBar(text: $viewModel.features)
                    }
```

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIBarProgressStyle.swift:
```diff
@@ -0,0 +1,38 @@
+        let progress = configuration.fractionCompleted ?? 0.0
+
+        GeometryReader { geometry in
+            VStack(alignment: .leading) {
```
Do we need the `VStack`? I think it can be safely removed.

pmusolino replied:
8421de758d8b410109ab6a3675a9d562b0e1051c

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIBarProgressStyle.swift:
```diff
@@ -0,0 +1,38 @@
+        GeometryReader { geometry in
+            VStack(alignment: .leading) {
+                RoundedRectangle(cornerRadius: Layout.radius)
+                    .fill(color.opacity(0.5))
```
Super nit: This can be moved to an `Constants` enum.

pmusolino replied:
e301da010ad2a7f7a7ab45dede4e178478901b2c

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBar.swift:
```diff
@@ -0,0 +1,46 @@
+            .animation(.easeIn, value: viewModel.status)
+            .padding(.top, Layout.padding/2)
+        })
+        .padding([.top, .bottom, .leading, .trailing], Layout.padding)
```
This can be simplified as below. 
```suggestion
        .padding(Layout.padding)
```

pmusolino replied:
Done 49533ded0b1c9647f7c60722c67bca4182a055bc

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBar.swift:
```diff
@@ -0,0 +1,46 @@
+            .padding(.top, Layout.padding/2)
+        })
+        .padding([.top, .bottom, .leading, .trailing], Layout.padding)
+        .background(Color(UIColor.listBackground))
```
Super nit:
```suggestion
        .background(Color(UIColor.listBackground))
```

pmusolino replied:
I think your suggestion looks the same as my code. Did you have in mind something different?

selanthiraiyan reviewed (commented):
Thanks for working on this, Paolo!  This looks good to me in general. 🎉 

I left a few nits/comments to make this better.

pmusolino committed:
Merge commit 'f0c730dedc32c031349d5fc837c9def54db6d93b' into issue/13118-progress-bar-and-label-with-dynamic-color-and-text

pmusolino committed:
fix: Indentation adjustment

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/213">19.5</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino committed:
fix: add manually empty space and fixed `amostDone` state

pmusolino committed:
fix: NSLocalizedString comment

pmusolino committed:
fix: improvement in almostDone state

pmusolino committed:
update: removed VStack

pmusolino committed:
update: removed padding

pmusolino committed:
Update WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/ProductCreationAIStartingInfoView.swift

Co-authored-by: Sharma Elanthiraiyan <sharma.elanthiraiyan@automattic.com>

pmusolino committed:
feat: added Constants

pmusolino commented:
@selanthiraiyan ready for another check. I will address all the design suggestions of @joe-keenan in another PR.

pmusolino committed:
fix: swiftlint errors

- WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:128:1: Line should be 163 characters or less; currently it has 164 characters (line_length)
- WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIBarProgressStyle.swift:31:1: Lines should not have trailing whitespace (trailing_whitespace)
- WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductCreationAIPromptProgressBarViewModelTests.swift:58:1: Line should be 163 characters or less; currently it has 203 characters (line_length)

pmusolino committed:
fix: compiling error

pmusolino committed:
fix: unit test

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:
```diff
@@ -0,0 +1,133 @@
+            case .start:
+                return ""
+            case .inProgress:
+                return NSLocalizedString(
```
Kindly consider moving the localized strings to a `Localization` enum. The same comments for other strings as well. https://github.com/woocommerce/woocommerce-ios/blob/trunk/docs/naming-conventions.md#string-constants-in-nested-enums

pmusolino replied:
👍 e9fa2f250d87945dd590c8a0a7e68b38d7909755

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:
```diff
@@ -0,0 +1,133 @@
+                    "productCreationAIPromptProgressBar.main.inProgress",
+                    value: "Add more details.",
+                    comment: "State when more details need to be added for the main prompt description suggestion in product creation with AI."
+                ) + " "
```
Opinion: Instead of appending empty space here, I would prefer adding an empty text from the calling place https://github.com/woocommerce/woocommerce-ios/pull/13280/files#diff-dfcce39268903acf8f3334776759bad7d7d85a4dbc0974b2eaecd6cbeef4f036R17. i.e. Add `Text(" ")` at the calling place.

pmusolino replied:
👍 028083b199b36183b96119f1dc55358d1ba70d60

selanthiraiyan commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Add Product/AddProductWithAI/ProductCreationAIPromptProgressBarViewModelTests.swift:
```diff
@@ -0,0 +1,141 @@
+        XCTAssertEqual(status.mainDescription, "")
+        XCTAssertNotNil(status.secondaryDescription)
+        XCTAssertEqual(status.secondaryDescription, NSLocalizedString(
+            "productCreationAIPromptProgressBar.secondary.start",
```
Nit: If we introduce a `Localization` nested enum, we can use it directly from the unit tests instead of duplicating the strings here. Applies to all tests here.

pmusolino replied:
Updated 👍 3fd7edb46fedf1e90ab9db8cc1b375cbeffae2fc

selanthiraiyan commented on WooCommerce/Classes/ViewRelated/Products/Add Product/AddProductWithAI/StartingInfo/PromptProgressBar/ProductCreationAIPromptProgressBarViewModel.swift:
```diff
@@ -0,0 +1,133 @@
+    private func updateStatusBasedOnText(_ text: String) {
+        let wordCount = text.split { $0 == " " || $0.isNewline }.count
+
+        switch wordCount {
+        case 0:
+            status = .start
+        case 1...10:
+            status = .inProgress
+        case 11...19:
+            status = .halfway
+        case 20...30:
+            status = .almostDone
+        default:
+            status = .completed
```
Opinion: Based on my testing, the `wordCount` we defined here looks like a strict expectation. Do you think we can reduce the expected word count? If it makes sense, consider starting a slack thread to discuss this further with the team. 

<img src="https://github.com/woocommerce/woocommerce-ios/assets/524475/737dc074-77b1-4942-810b-57a65bf8fe93" width="320"/>

JorgeMucientes replied:
I don’t have a strong opinion tbh. I think is tricky to come up with good numbers that will work well for all languages.  But yeah, I think overall we can maybe reduce all the ranges by 2-3 words?

pmusolino replied:
Discussion moved here C03L1NF1EA3/p1720706347936069

pmusolino replied:
Merging this now, I can update the ranges later on based on the results of the discussion.

selanthiraiyan reviewed (approved):
This works as described, Paolo! 🚢 

I am adding a few nitpicks and suggestions.

pmusolino committed:
fix: spacing

pmusolino committed:
feat: Localization enum

pmusolino committed:
feat: added space in the calling place

pmusolino committed:
feat: used Localized string directly in unit tests
